src/transfer-manager.ts

```diff
@@ -369,6 +575,98 @@ export class TransferManager {
+   * @property {string} [uploadId] If specified attempts to resume a previous upload.
+   * @property {Map} [partsMap] If specified alongside uploadId, attempts to resume a previous upload from the last chunk
+   * specified in partsMap
+   * @experimental
+   */
+  /**
+   * Upload a large file in chunks utilizing parallel upload opertions. If the upload fails, an uploadId and
+   * map containing all the successfully uploaded parts will be returned to the caller. These arguments can be used to
+   * resume the upload.
+   *
+   * @param {string} [filePath] The path of the file to be uploaded
+   * @param {UploadFileInChunksOptions} [options] Configuration options.
+   * @param {MultiPartHelperGenerator} [generator] A function that will return a type that implements the MPU interface. Most users will not need to use this.
+   * @returns {Promise<void>} If successful a promise resolving to void, otherwise a error containing the message, uploadid, and parts map.
+   *
+   * @example
+   * ```
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const bucket = storage.bucket('my-bucket');
+   * const transferManager = new TransferManager(bucket);
+   *
+   * //-
+   * // Upload a large file in chunks utilizing parallel operations.
+   * //-
+   * const response = await transferManager.uploadFileInChunks('large-file.txt');
+   * // Your bucket now contains:
+   * // - "large-file.txt"
+   * ```
+   *
+   * @experimental
+   */
+  async uploadFileInChunks(
+    filePath: string,
+    options: UploadFileInChunksOptions = {},
+    generator: MultiPartHelperGenerator = defaultMultPartGenerator
+  ): Promise<void> {
+    const chunkSize =
+      options.chunkSizeBytes || UPLOAD_IN_CHUNKS_DEFAULT_CHUNK_SIZE;
+    const limit = pLimit(
+      options.concurrencyLimit || DEFAULT_PARALLEL_CHUNKED_UPLOAD_LIMIT
+    );
+    const maxQueueSize =
+      options.maxQueueSize ||
+      options.concurrencyLimit ||
+      DEFAULT_PARALLEL_CHUNKED_UPLOAD_LIMIT;
+    const fileName = options.uploadName || path.basename(filePath);
+    const mpuHelper = generator(
+      this.bucket,
+      fileName,
+      options.uploadId,
+      options.partsMap
+    );
+    let partNumber = 1;
+    let promises = [];
+    try {
+      if (options.uploadId === undefined) {
+        await mpuHelper.initiateUpload();
+      }
+      const startOrResumptionByte = mpuHelper.partsMap!.size * chunkSize;
+      const readStream = createReadStream(filePath, {
+        highWaterMark: chunkSize,
+        start: startOrResumptionByte,
+      });
+      for await (const curChunk of readStream) {
+        if (promises.length >= maxQueueSize) {
+          await Promise.all(promises);
+          promises = [];
+        }
+        promises.push(
+          limit(() => mpuHelper.uploadPart(partNumber++, curChunk))
+        );
+      }
+      await Promise.all(promises);
+      await mpuHelper.completeUpload();
```

Comment:
What's the return statement here?

Reply:
The return is of type `Promise<void>` in the success case there is no return value to the user. If something goes wrong we throw an exception with the error, uploadId, and mapping of completed parts that were uploaded. This allows the user to call again supplying the uploadId and parts so that we may resume from where the failure occurred.

Reply:
I change it to return the full response just to be inline with the raw API.

Reply:
Now returns `Promise<GaxiosResponse | undefined>`. The GaxiosResponse is the full API response.